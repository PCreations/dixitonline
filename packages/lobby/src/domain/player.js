export const makePlayer = ({ id, name } = {}) => {
  if (!id) throw new Error('Player must contain an id');
  if (!name) throw new Error('Player must have a name');
  return { id, name };
};

export const equals = (player1, player2) => player1.id === player2.id;
