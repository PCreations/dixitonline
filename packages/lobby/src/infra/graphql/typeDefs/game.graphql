type LobbyGame {
  id: ID!
  host: LobbyPlayer!
  players: [LobbyPlayer!]!
}
