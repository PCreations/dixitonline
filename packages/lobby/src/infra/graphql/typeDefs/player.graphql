type LobbyPlayer {
  id: ID!
  name: String!
}
