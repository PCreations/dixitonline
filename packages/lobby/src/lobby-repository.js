import { v1 as uuidv1 } from 'uuid';

export const makeLobbyRepository = ({ uuid = uuidv1, firestore = makeNullFirestore() } = {}) => {
  const lobbyGames = firestore.collection('lobby-games');
  return {
    getNextGameId() {
      return uuid();
    },
    createGame(game) {
      return lobbyGames.doc(game.id).set(game);
    },
    async getGameById(id) {
      const doc = await lobbyGames.doc(id).get();
      return doc.data();
    },
  };
};

export const makeNullLobbyRepository = ({ nextGameId, gamesData }) =>
  makeLobbyRepository({ uuid: nullUuid(nextGameId), firestore: makeNullFirestore(gamesData) });

const nullUuid = expectedUuid => () => expectedUuid;

const makeNullFirestore = (gamesInitialData = {}) => {
  const gamesData = { ...gamesInitialData };
  return {
    collection() {
      return {
        doc(gameId) {
          return {
            get() {
              return {
                data() {
                  return gamesData[gameId];
                },
              };
            },
          };
        },
      };
    },
  };
};
