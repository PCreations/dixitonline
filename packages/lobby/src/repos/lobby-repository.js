import { v1 as uuidv1 } from 'uuid';

export const makeLobbyRepository = ({ uuid = uuidv1, firestore = makeNullFirestore() } = {}) => {
  const lobbyGames = firestore.collection('lobby-games');
  return {
    getNextGameId() {
      return uuid();
    },
    createGame(game) {
      return lobbyGames.doc(game.id).set(game);
    },
    async getGameById(id) {
      const doc = await lobbyGames.doc(id).get();
      return doc.data();
    },
    getAllGames() {
      return lobbyGames.get().then(snapshot => {
        const games = [];
        snapshot.forEach(doc => games.push(doc.data()));
        return games;
      });
    },
  };
};

export const makeNullLobbyRepository = ({ nextGameId, gamesData }) =>
  makeLobbyRepository({ uuid: nullUuid(nextGameId), firestore: makeNullFirestore(gamesData) });

const nullUuid = expectedUuid => () => expectedUuid;

const makeNullFirestore = (gamesInitialData = {}) => {
  const gamesData = { ...gamesInitialData };
  return {
    collection() {
      return {
        doc(gameId) {
          return {
            get() {
              return {
                data() {
                  return gamesData[gameId];
                },
              };
            },
            set(game) {
              gamesData[gameId] = game;
            },
          };
        },
        async get() {
          const docs = Object.values(gamesData).map(game => ({
            data() {
              return game;
            },
          }));
          return {
            forEach: docs.forEach.bind(docs),
          };
        },
      };
    },
  };
};
