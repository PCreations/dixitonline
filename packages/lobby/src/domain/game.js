import { playerJoinedGame } from './events';
import { equals as playerEquals } from './player';

export class GameAlreadyJoinedByPlayerError extends Error {}

export const makeGame = ({ id, host, players = [] } = {}) => {
  if (!id) throw new Error('Game must contain an id');
  if (!host) throw new Error('Game must have an host');

  return Object.freeze({
    id,
    host,
    players,
  });
};

const getAllPlayers = game => [game.host, ...game.players];

export const joinPlayer = (game, player) => {
  if (getAllPlayers(game).some(playerEquals.bind(null, player))) throw new GameAlreadyJoinedByPlayerError();
  return [
    makeGame({ ...game, players: game.players.concat(player) }),
    [playerJoinedGame({ gameId: game.id, playerId: player.id })],
  ];
};
