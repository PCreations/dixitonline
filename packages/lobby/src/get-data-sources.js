import { makeLobbyRepository } from './lobby-repository';

export const makeGetDataSources = ({ lobbyRepository = makeLobbyRepository() } = {}) => () => ({
  lobbyRepository,
});
