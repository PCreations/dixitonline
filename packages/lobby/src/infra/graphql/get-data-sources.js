import { makeLobbyRepository } from '../../repos/lobby-repository';

export const makeGetDataSources = ({ lobbyRepository = makeLobbyRepository() } = {}) => () => ({
  lobbyRepository,
});
