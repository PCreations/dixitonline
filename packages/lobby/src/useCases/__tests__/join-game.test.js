import gql from 'graphql-tag';
import { createTestClient } from 'apollo-server-testing';
import { makeNullLobbyRepository } from '../../repos/lobby-repository';
import { makeTestServer } from '../../__tests__/test-server';
import { buildLobbyRepositoryInitialGames } from '../../__tests__/dataBuilders/lobby-repository-initial-games';
import { buildTestGame } from '../../__tests__/dataBuilders/game';
import { buildTestPlayer } from '../../__tests__/dataBuilders/player';
import { makeGetDataSources } from '../../infra/graphql/get-data-sources';

describe('join game', () => {
  test('a player can join a game', async () => {
    // arrange
    const host = buildTestPlayer()
      .withId('p1')
      .withName('player1')
      .build();
    const game = buildTestGame()
      .withId('g1')
      .withHost(host)
      .build();
    const intialGames = buildLobbyRepositoryInitialGames()
      .withGames([game])
      .build();

    const lobbyRepository = makeNullLobbyRepository({ intialGames });
    const dispatchDomainEvents = jest.fn();
    const server = makeTestServer({
      getDataSources: makeGetDataSources({
        lobbyRepository,
      }),
      dispatchDomainEvents,
      currentUserId: 'p2',
      currentUserUsername: 'player2',
    });
    const LOBBY_JOIN_GAME = gql`
      mutation LobbyCreateGame($lobbyJoinGameInput: LobbyJoinGameInput!) {
        lobbyJoinGame(lobbyJoinGameInput: $lobbyJoinGameInput) {
          game {
            id
            host {
              id
              name
            }
            players {
              id
              name
            }
          }
        }
      }
    `;

    // act
    const { mutate } = createTestClient(server);
    const response = await mutate({
      mutation: LOBBY_JOIN_GAME,
      variables: { lobbyJoinGameInput: { gameId: 'g1' } },
      operationName: 'LobbyCreateGame',
    });

    // assert
    expect(response).toMatchSnapshot();
    expect(dispatchDomainEvents).toHaveBeenCalledWith([
      { type: '[lobby] - a new player has joined a game', payload: { playerId: 'p2', gameId: 'g1' } },
    ]);
  });
});
