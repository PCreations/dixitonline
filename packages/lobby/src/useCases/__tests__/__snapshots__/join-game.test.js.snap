// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`join game a player can join a game 1`] = `
Object {
  "data": Object {
    "lobbyJoinGame": Object {
      "game": Object {
        "host": Object {
          "id": "p1",
          "name": "player1",
        },
        "id": "g1",
        "players": Array [
          Object {
            "id": "p2",
            "name": "player2",
          },
        ],
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
